src/DependencyInjection/DI/src/ServiceLookup/ServiceProviderEngineScope.cs

```diff
@@ -127,7 +127,10 @@ private List<object> BeginDispose()
                 toDispose = _disposables;
                 _disposables = null;
 
-                ResolvedServices.Clear();
+                // Not clearing ResolvedServices here because there might be a compilation running in background
+                // trying to get a cached singleton service instance and if it won't find 
+                // it it will try to create a new one tripping the Debug.Assert in CaptureDisposable
+                // and leaking a Disposable object in Release mode
```

Comment:
What should happen in CaptureDisposable if the container is already disposed? Shouldn't we throw an ODE that can be observed by the top-level GetService() call?

Reply:
It shouldn't get called after dispose. We have ODE checks on all customer accessed entry points. This one is internal that that's why it was able to slip past those checks.